### ISSUE TITLE
DSPARK: draft worker inherits speculative_num_draft_tokens (gamma+1) into its attention backend while the proposer emits gamma rows — OOB KV reads with the triton draft backend

### ISSUE THREAD

**Issue by abatilo** [sgl-project/sglang#30555]
### Summary

On the `sglang-dspark` branch (692c5f7d, PR #30261), serving a DSPARK draft via a separate `--speculative-draft-model-path` with `--speculative-draft-attention-backend triton` produces out-of-bounds KV reads on the very first propose step. The draft `TpModelWorker` is built from a deep copy of the target `ServerArgs`, so its attention backend inherits `speculative_num_draft_tokens = gamma + 1` (the TARGET verify window), while the DSPARK proposer only emits `gamma` query rows per sequence.

### Repro shape

- Target: a DeepSeek-class sparse model, TP8, `--speculative-algorithm DSPARK --speculative-dspark-block-size 7`
- Draft: dense 6-layer `Qwen3DSparkModel` (head_dim 96, markov_rank 256, trained standalone), loaded from a local dir via `--speculative-draft-model-path`
- `--speculative-draft-attention-backend triton` (flashinfer rejects head_dim 96 in `VariableLengthMergeStates`)
- Image `lmsysorg/sglang:dev-dspark`

First `/v1/chat/completions` request → all TP ranks spam `Indexing.cu:1515 indexSelectSmallIndex: Assertion 'srcIndex < srcSelectDimSize' failed` and the scheduler dies.

### Mechanism

1. `triton_backend.py:150`: `self.num_draft_tokens = model_runner.server_args.speculative_num_draft_tokens`. For the draft worker this is the deep-copied target value, i.e. `gamma + 1 = 8`.
2. The DSPARK proposer (`dspark_draft_proposer.py`) builds `draft_block_ids` as `(bs, gamma)` and runs the draft forward in `ForwardMode.TARGET_VERIFY` with `bs * gamma = bs * 7` query rows.
3. In the triton backend's target-verify metadata, `qo_indptr = arange(0, (1+bs) * num_draft_tokens, step=num_draft_tokens)` strides by 8 over 7-rows-per-sequence tensors. Every sequence after the first reads shifted rows and the tail indexes past the real data → attention consumes unwritten KV slots.
4. Downstream visible effect: draft hidden states come back all-zero (fp8 KV pool, zero-initialized) or all-NaN (bf16 pool, uninitialized memory). NaN logits make the step sampler emit `2147483647` (INT32_MAX), which then trips the device assert inside the replicated markov embedding `nn.Embedding(vocab, markov_rank)`.

Two pieces of evidence that pinned it:

- Kernel fingerprint: `indexSelectSmallIndex` launches `ceil(sliceSize/128)` blocks of 128 threads; the observed `blocks [0..1]` matches slice = markov_rank 256 exactly (and the assert spams identically on all ranks because the markov embedding is replicated, not TP-sharded).
- Flipping `--kv-cache-dtype fp8_e4m3` → `auto` turned the all-zero draft hidden into all-NaN — i.e., the reads land in unwritten pool memory, not quantization corruption.

### Fix that we validated

Pinning the draft worker's value in `build_draft_tp_worker` (`draft_worker_common.py`) makes the server healthy end-to-end (single and concurrent requests, long generations):

```python
draft_server_args.context_length = target_model_config.context_len
if getattr(server_args, "speculative_dspark_block_size", None):
    draft_server_args.speculative_num_draft_tokens = int(
        server_args.speculative_dspark_block_size
    )
```

The target worker keeps its own backend instance with `gamma + 1`, which is correct for verify.

### Follow-on: decode CUDA-graph capture

With the fix above, decode CUDA-graph capture still fails: the decode graph runner tiers token counts by the global `gamma + 1` (e.g. 1536 = 192 x 8) while the dspark sampler views by `gamma`:

```
Exception: Capture cuda graph failed: shape '[219, 7, -1]' is invalid for input of size 237895680
```

(237895680 / vocab = 1536 rows; 1536 / 7 = 219.4). So `--cuda-graph-backend-decode=disabled` is currently required for this configuration. Aligning the graph runner's spec token multiplier with the draft's gamma would presumably fix that too.

The bundled-checkpoint paths (e.g. DeepSeek-V4-Flash-DSpark, which owns `forward_embed` and doesn't take the triton draft backend) don't hit this, which is probably why it survived — the separate-draft + triton path just seems untested.

Related: DSpark roadmap #30344, the 2026-07-06 DSpark blog post.
